olive/passes/pytorch/gptq.py

```diff
@@ -0,0 +1,219 @@
+            or not isinstance(dataset[0], dict)
+            or ("input_ids" not in dataset[0] or "attention_mask" not in dataset[0])
+        ):
+            raise ValueError(
+                "Provided dataset is invalid. The returned datasets is a list of tokenized data "
+                "(e.g. [{ 'input_ids': [ 1, 100, 15, ... ],'attention_mask': [ 1, 1, 1, ... ]},...])"
+            )
+
+        pytorch_model = model.load_model()
+        quantize_config = BaseQuantizeConfig(
+            bits=config["bits"],
+            group_size=config["group_size"],
+            damp_percent=config["damp_percent"],
+            static_groups=config["static_groups"],
+            true_sequential=config["true_sequential"],
+            desc_act=config["desc_act"],
+            sym=config["sym"],
+        )
+
+        def get_onnx_quant_linear(*args, **kwargs):
+            return QuantLinearORT
+
+        if hasattr(pytorch_model, "config") and pytorch_model.config.model_type in GPTQ_CAUSAL_LM_MODEL_MAP:
+            model_type = pytorch_model.config.model_type
+            model_class = GPTQ_CAUSAL_LM_MODEL_MAP[model_type]
+            quantized_model = model_class(pytorch_model, False, quantize_config)
+        else:
+            quantized_model = BaseGPTQForCausalLM(pytorch_model, False, quantize_config)
+            if not (config["layers_block_name"] and config["outside_layer_modules"] and config["inside_layer_modules"]):
+                raise ValueError(
+                    "Can't get layers_block_name to quantize automatically, "
+                    "please set layers_block_name, outside_layer_modules and inside_layer_modules in config."
+                )
+            quantized_model.layers_block_name = config["layers_block_name"]
+            quantized_model.outside_layer_modules = config["outside_layer_modules"]
+            quantized_model.inside_layer_modules = config["inside_layer_modules"]
+
+        import auto_gptq
+
+        original = auto_gptq.modeling._utils.dynamically_import_QuantLinear  # pylint: disable=protected-access
+        try:
+            # Replace QuantLinear in autogptq with QuantLinearORT for quant linear layer packing
+            auto_gptq.modeling._utils.dynamically_import_QuantLinear = (  # pylint: disable=protected-access
+                get_onnx_quant_linear
+            )
+
+            # Autogpq quantize_model currently only support cuda device. It accepts model on cpu but
+            # will move each block(layer) to cuda before quantization and move back to cpu when finished.
+            quantized_model.quantize(dataset)
+        finally:
+            auto_gptq.modeling._utils.dynamically_import_QuantLinear = original  # pylint: disable=protected-access
+
+        quantized_model = quantized_model.model
+        assert self.accelerator_spec.accelerator_type == Device.GPU
+
+        output_model_path = Path(output_model_path).with_suffix(".pt")
+        torch.save(quantized_model, output_model_path)
```

Comment:
This looks good to me. Just to be sure, the model when loading needs olive installed since it has modules that are defined in olive? 
Not an issue since we expect to convert the model to onnx after this pass anyways.